### ISSUE TITLE
indexer backward: dk (index_k grad) = garbage (8e33) in full miles-indexer integration; isolation always clean

### ISSUE THREAD

**Issue by zhshgmail** [tile-ai/tilelang-mlir-ascend#99]
## What

`examples/deepseek_v4/example_miles_indexer_integration.py` produces a **garbage `dk` (index_k gradient)** while `dq`/`dw` are correct.

```
dk kernel max abs (finite): 8.23e+33      (CPU ref max abs: 0.5988)
max abs err vs CPU ref:  dk: 8.23e+33      (dq: 0.00114, dw: 0.00000)
```

Localized to the backward per-seq loop iteration where all `topk` grad-score slots are active (`grad_scores=[1,1,1,1]`, `idx=[3,2,0,1]`, fully-populated). Earlier seq positions (with trailing zero grad-score slots) are clean.

## Reproduction

Run `python examples/deepseek_v4/example_miles_indexer_integration.py` on Ascend (A3 / Ascend910_9382, current `tilelang-mlir-ascend` + bundled bishengir). The `=== max abs err vs CPU ref ===` block reports `dk: 8.2e+33`.

## What we ruled out (isolation tests — all CLEAN, only the full integration reproduces)

| Hypothesis | Test | Result |
|---|---|---|
| index ordering | standalone bwd with idx `[3,2,0,1]` (vs default `arange`) | CLEAN |
| specific input values | dumped the exact garbage-producing inputs (q/k/w/idx/grad_scores), fed to a fresh standalone `lighting_indexer_bwd` | CLEAN (0.43, scales linearly) |
| repeated invocation | same kernel object called 6× in a loop | CLEAN |
| async / stream ordering | `torch.npu.synchronize()` before each bwd call | still garbage |
| prior backward calls | skip s=0/1/2 bwd, run only the failing s=3 bwd (forward phase intact) | still garbage |
| forward tilelang kernel exec | replace `lighting_indexer_fwd` call with pure-torch logits | still garbage |
| forward kernel build coexistence | remove the `lighting_indexer_fwd(...)` build entirely | still garbage |
| active-slot progression | same kernel object, exact s=0..3 grad/idx progression | CLEAN |

So: every controlled isolation is clean; **only the full forward+topk+backward integration script reproduces it.** The remaining difference between isolation and integration is the real tensor lifecycle / NPU memory layout in the integration (`ctx.save_for_backward` tensors, autograd graph, `.contiguous()` slicing of the batched tensors, the NPU memory allocation pattern left by the full forward). The `dk` path uses `T.atomic_addx4(dIndexK[cur_idx, ...])` scatter, which is the prime suspect for residual / aliased NPU state, but the same scatter is clean in isolation — pointing at cross-kernel NPU memory state rather than the bwd kernel logic.

This likely needs an NPU memory-allocator trace / UB-residue inspection to pin down — beyond what we can do from the Python/IR layer.

Environment: Ascend A3 (Ascend910_9382), bfloat16 q/k, fp32 grads, `lighting_indexer_bwd(seq_len=1, block_I=min(8,topk))`.
